routing/probe_test.go

```diff
@@ -0,0 +1,822 @@
+	require.NoError(t, err)
+
+	// after adding first node the probe should be idle since the
+	// connectivity check will be scheduled for the future
+	state := sm.Advance(ctx, &EventProbeAdd[key.Key8]{
+		NodeID: kadtest.NewID(key.Key8(4)),
+	})
+	require.IsType(t, &StateProbeIdle{}, state)
+
+	// remains idle
+	state = sm.Advance(ctx, &EventProbePoll{})
+	require.IsType(t, &StateProbeIdle{}, state)
+
+	// advance time by one revisit interval
+	clk.Add(cfg.CheckInterval)
+	state = sm.Advance(ctx, &EventProbePoll{})
+	require.IsType(t, &StateProbeConnectivityCheck[key.Key8, kadtest.StrAddr]{}, state)
+
+	// the probe state machine should attempt to contact the next node
+	st := state.(*StateProbeConnectivityCheck[key.Key8, kadtest.StrAddr])
+
+	// the connectivity check should be for the right node
+	require.True(t, key.Equal(key.Key8(4), st.NodeInfo.ID().Key()))
+
+	state = sm.Advance(ctx, &EventProbePoll{})
+	require.IsType(t, &StateProbeWaitingWithCapacity{}, state)
+}
+
+func TestProbeRemoveDeletesNodeValue(t *testing.T) {
+	ctx := context.Background()
+	clk := clock.NewMock()
+
+	cfg := DefaultProbeConfig()
+	cfg.Clock = clk
+	cfg.CheckInterval = 10 * time.Minute
+
+	// Set concurrency to allow more than one check to run
+	cfg.Concurrency = 2
+
+	rt := simplert.New[key.Key8, kad.NodeID[key.Key8]](kadtest.NewID(key.Key8(128)), 5)
+	rt.AddNode(kadtest.NewID(key.Key8(4)))
+
+	sm, err := NewProbe[key.Key8, kadtest.StrAddr](rt, cfg)
+	require.NoError(t, err)
+
+	// after adding first node the probe should be idle since the
+	// connectivity check will be scheduled for the future
+	state := sm.Advance(ctx, &EventProbeAdd[key.Key8]{
+		NodeID: kadtest.NewID(key.Key8(4)),
+	})
+	require.IsType(t, &StateProbeIdle{}, state)
+
+	// remove the node
+	state = sm.Advance(ctx, &EventProbeRemove[key.Key8]{
+		NodeID: kadtest.NewID(key.Key8(4)),
+	})
+
+	// state should remain idle
+	require.IsType(t, &StateProbeIdle{}, state)
+
+	// advance time by one revisit interval
+	clk.Add(cfg.CheckInterval)
+
+	// state remains idle since there are no nodes to probe
+	state = sm.Advance(ctx, &EventProbePoll{})
+	require.IsType(t, &StateProbeIdle{}, state)
+}
+
+func TestNodeValueList(t *testing.T) {
+	t.Run("put new", func(t *testing.T) {
+		t.Parallel()
+
+		clk := clock.NewMock()
+		l := NewNodeValueList[key.Key8]()
+		nv := &nodeValue[key.Key8]{
+			NodeID:       kadtest.NewID(key.Key8(4)),
+			NextCheckDue: clk.Now(),
+		}
+
+		l.Put(nv)
+
+		got, found := l.Get(kadtest.NewID(key.Key8(4)))
+		require.True(t, found)
+		require.True(t, key.Equal(got.NodeID.Key(), key.Key8(4)))
+	})
+
+	t.Run("put replace", func(t *testing.T) {
+		t.Parallel()
+
+		clk := clock.NewMock()
+		l := NewNodeValueList[key.Key8]()
+		nv1 := &nodeValue[key.Key8]{
+			NodeID:       kadtest.NewID(key.Key8(4)),
+			NextCheckDue: clk.Now(),
+		}
+
+		l.Put(nv1)
+
+		nv2 := &nodeValue[key.Key8]{
+			NodeID:       kadtest.NewID(key.Key8(4)),
+			NextCheckDue: clk.Now().Add(-time.Minute),
```

Comment:
Might be worth enforcing that even if `nv2.NextCheckDue` is after `nv1.NextCheckDue` the value will be replaced. So that we're not clever in that regard.

Reply:
Done